File: Stack/Min_Stack/3_Optimal.java
import java.util.Stack;

public class MinStack {
    private Stack<Long> stack;
    private long min;
    public MinStack() {
        stack = new Stack<>();
    }
    public void push(int x) {
        if(stack.isEmpty()) {
            stack.push((long)x);
            min = x;
        } else if(x >= min) {
            stack.push((long)x);
        } else {
            // encode previous min
            stack.push(2L * x - min);
            min = x;
        }
    }
    public void pop() {
        if(stack.isEmpty()) return;
        long top = stack.pop();
        if(top < min) {
            // retrieve previous min
            min = 2L * min - top;
        }
    }
    public int top() {
        long top = stack.peek();
        if(top >= min) return (int)top;
        else return (int)min;
    }
    public int getMin() {
        return (int)min;
    }
}

File: Stack/Min_Stack/3_Optimal.cpp
#include <stack>
using namespace std;

class MinStack {
    stack<long long> st;
    long long minVal;
public:
    MinStack() {}
    void push(int x) {
        if(st.empty()) {
            st.push(x);
            minVal = x;
        } else if(x >= minVal) {
            st.push(x);
        } else {
            st.push(2LL * x - minVal); // encoded value
            minVal = x;
        }
    }
    void pop() {
        if(st.empty()) return;
        long long top = st.top();
        st.pop();
        if(top < minVal) {
            // retrieve previous minimum
            minVal = 2LL * minVal - top;
        }
    }
    int top() {
        long long top = st.top();
        if(top >= minVal) return (int)top;
        else return (int)minVal;
    }
    int getMin() { return (int)minVal; }
};
